blog/_posts/2019-09-06-profilers.md

```diff
@@ -0,0 +1,405 @@
+
+In addition, the `pprof` tool has a sandwich text report that summarizes the caller and callee list around each profile sample:
+
+```
+      Type: events
+Showing nodes accounting for 632, 100% of 632 total
+----------------------------------------------------------+-------------
+      flat  flat%   sum%        cum   cum%   calls calls% + context
+----------------------------------------------------------+-------------
+                                               273 52.20% |   fib REPL:-1
+                                               273 52.20% |   fib_r REPL:1
+       417 65.98% 65.98%        523 82.75%                | fib REPL:2
+                                               273 52.20% |   fib_r REPL:1
+----------------------------------------------------------+-------------
+                                               273 72.03% |   fib REPL:2
+       178 28.16% 94.15%        379 59.97%                | fib_r REPL:1
+                                               273 72.03% |   fib REPL:2
+                                                33  8.71% |   + ./int.jl:2
+                                                 1  0.26% |   - ./int.jl:2
+----------------------------------------------------------+-------------
+                                                33   100% |   + ./int.jl:2 (inline)
+        33  5.22% 99.37%         33  5.22%                | + ./int.jl:53
+----------------------------------------------------------+-------------
+                                                 1   100% |   - ./int.jl:2 (inline)
+         1  0.16% 99.53%          1  0.16%                | - ./int.jl:52
+----------------------------------------------------------+-------------
+```
+
+Let's combine all those ideas to make a new option for our Julia plain-text tree view: flattened recursion.
+
+  * recur – Controls the recursion handling in `:tree` format. `off` (default) prints the tree as normal. `flat` instead compresses any recursion (by ip), showing the approximate effect of converting any self-recursion into an iterator. `flatc` does the same but also includes collapsing of C frames (may do odd things around `jl_apply`).
+
+This is no longer a literal interpretation of the stack, but shows a particular linearization where each recursive call was re-written as a loop.
+
+```
+julia> Profile.print(format=:tree, recur=:flat)
+106 ./REPL[1]:1; fib_r(::Int64)
+250 ./REPL[2]:2; fib(::Int64)
+275 ./task.jl:333; (::REPL.var"##26#27"{REPL.REPLBackend})()
+ 275 /Users/jameson/julia/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:118; macro expansion
+  275 /Users/jameson/julia/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86; eval_user_input(::Any, ::REPL.REPLBackend)
+   275 ./boot.jl:330; eval(::Module, ::Any)
+  ╎ 273 ./REPL[2]:2; fib(::Int64)
+  ╎  309 ./REPL[1]:1; fib_r(::Int64)
+  ╎   469 ./REPL[2]:2; fib(::Int64)
```

Comment:
469 > 309 seems a bit counterintuitive.

Reply:
yes it does. I'll review my code and see if something isn't being updated correctly. (It wouldn't be the first bug that writing out carefully the inputs and outputs here has caught! I've just already squashed the others into the history, haha.)

Reply:
Ah, haha, this is the same mistake I pointed out in Speedscope above (https://github.com/JuliaLang/www.julialang.org/pull/447#discussion_r322006957). I need to rework some things.